src/types/Expr.d.ts

```diff
@@ -1,5 +1,118 @@
-export default class Expr {
+import { values } from './values'
+
+export default Expr
+
+export class Expr<T = any> {
   constructor(obj: object)
 
   static toString(expr: Expr): string
+
+  // This prevents structural type equality with empty objects.
+  private _exprType: T
+}
+
+/** Materialize an `Expr` type into its result type. */
+export type Materialize<T> = T extends Expr<infer U>
+  ? U
+  : T extends ReadonlyArray<infer U>
+  ? (U[] extends T
+      ? ReadonlyArray<Materialize<U>>
+      : { [P in keyof T]: Materialize<T[P]> })
+  : T
+
+/** Add support for `Expr` types to any type. */
+export type ExprVal<T = unknown> =
+  | Expr<Exclude<T, Expr>>
+  | (T extends Expr | Lambda
+      ? T
+      : T extends ReadonlyArray<infer U>
+      ? (U[] extends T
+          ? ReadonlyArray<ExprVal<U>>
+          : { [P in keyof T]: ExprVal<T[P]> })
+      : T extends object
+      ? { [P in keyof T]: ExprVal<T[P]> }
+      : T)
+
+export type ToExpr<T> = T extends Expr ? T : Expr<T>
+
+export type Lambda<In extends any[] = any[], Out = any> = (
+  ...args: { [P in keyof In]: ToExpr<In[P]> }
+) => Expr<Out>
+
+export interface Collection<T extends object> {
+  ref: Expr.CollectionRef<T>
+  ts: number
+  name: string
+  data?: object
```

Comment:
you can add a second parameter to that interface to represent the metadata to enforce this type instead of use `object`

Reply:
yup, i'm about to push such a change to support that :)